### ISSUE TITLE
Export PatternOptions and VariantGroups types?

### ISSUE THREAD

**Issue by igor10k** [vanilla-extract-css/vanilla-extract#694]
Could `PatternOptions` and `VariantGroups` be exported in `@vanilla-extract/recipes` to allow creating extendable recipes?

**Example:**

```ts
const createRecipe = <T extends VariantGroups>(
  extension: PatternOptions<T | typeof baseVariants>,
) => {
  return recipe({
    base: extension.base,
    variants: {
      ...baseVariants,
      ...extension.variants,
    },
    compoundVariants: extension.compoundVariants,
    defaultVariants: {
      ...baseDefaultVariants,
      ...extension.defaultVariants,
    },
  });
};

const button = createRecipe<typeof buttonVariants>({
  variants: buttonVariants,
  compoundVariants: [
    // mixing variants from baseVariants and buttonVariants here
  ]
});
```

There are ways to make it work without them but at the expense of losing auto-suggestions.

**Comment by graup** [comment]
You could always pull the types out with TS, for example:

```
import { recipe } from '@vanilla-extract/recipes';

type VariantGroups = NonNullable<Parameters<typeof recipe>[0]['variants']>;
type PatternOptions = Parameters<typeof recipe>[0];
```

Does this help?

**Comment by igor10k** [comment]
This works but is there any reason not to export them?

**Comment by askoufis** [comment]
> This works but is there any reason not to export them?

Quoting [a reply on discord](https://discordapp.com/channels/885877446098964512/885877446891667529/1080951947252731904) to a similar question:

> **michaletaranto**: As a general rule, we try to minimise the API footprint as it makes us less vulnerable to breaking changes in the future

**Comment by igor10k** [comment]
I see. Thank for the answers.

**Comment by gabro** [comment]
Hi all, following up on this issue (I can open a new one if preferred), we tried a similar workaround to the one proposed by @graup, but it introduced a huge performance regression when replacing `SprinklesProperties` with 

```ts
type VarargParameters<T extends (args: any) => any> = T extends (args: infer P) => any ? P : never;
type SprinklesProperties = VarargParameters<typeof createSprinkles>;
```

I've ran a CPU profile on the library using this, and it is very clearly pointing at `VarargParameters`.

The difference is ~5 seconds (with `SprinklesProperties`) vs ~2 minutes using the workaround.

Does anyone else face similar issues? I think the culprit here is that `Parameters` doesn't work with varargs arguments, and it's apparently very slow to roll your own manually.

I'm open to suggestions, but this is currently a blocker for us to upgrade to any newer version of Sprinkles.
